src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -330,15 +330,15 @@ take n (Fold fstep finitial fextract) = Parser step initial extract
 
     where
 
-    initial = Tuple' 0 <$> finitial
+    initial = Tuple' n <$> finitial
 
-    step (Tuple' i r) a = do
+    step (Tuple' i r) a
+      | i == 0 = Done 1 <$> fextract r
```

Comment:
We can write the code to use `Done 0` instead of `Done 1` so that we can return the result in the earlier step itself and we do not have to return an element back to the stream. The previous code in fact was doing that. We just need to replace `res` with `s1` in the `Done` constructor in the earlier code to fix the issue.

Reply:
Sorry, scratch the previous fix suggestion. How about this, with minimal changes to the old code and keeping the common branch case first:

```
    step (Tuple' i r) a
        | i < n = do
            res <- fstep r a
            let i1 = i + 1
                s1 = Tuple' i1 res
            if i1 < n
            then return $ Partial 0 s1
            else Done 0 <$> fextract res
        | otherwise = Done 1 <$> fextract r
```

Reply:
Also the current fix has a bug when `n` is negative, it will never terminate.

Can we also add these two test cases:

1) when n is 0
2) when n is negative